pilot/pkg/xds/ads.go

```diff
@@ -340,16 +371,26 @@ func (s *DiscoveryServer) shouldRespond(con *Connection, request *discovery.Disc
 	con.proxy.WatchedResources[request.TypeUrl].LastRequest = request
 	con.proxy.Unlock()
 
+	// Signal waiting semaphore in push operations that a valid
+	// ACK has been processed as determined by matched nonce
+	if features.EnableFlowControl {
+		select {
+		case <-con.semaphore:
+		case <-con.exitReceive:
```

Comment:
You send to this channel when push error, but when pushXds errored, envoy would not send ack/nack. So it will not run into this function

Reply:
The next time pilot needs push xds, it will block forever

Reply:
agree - this approach results in deadlock.